test_env_setup_util/env_setup.py

```diff
@@ -161,19 +167,36 @@ def _replace_variables(self, contents):
     def run(self):
         exit_code = ExitCode.Success
         results = {}
-        actions, actions_sources = self._load_env_setup_file(self._root_yaml)
-        actions = self._replace_variables(actions)
+        raw_actions, actions_sources = self._load_env_setup_file(
+            self._root_yaml
+        )
+        rendered_actions = self._replace_variables(raw_actions)
 
-        for idx, action in enumerate(actions, start=1):
+        try:
+            # Re-validate after replacing variables to ensure correctness
+            updated_actions = {"actions": rendered_actions}
+            validated_data = EnvSetup(**updated_actions)
```

Comment:
I think you can use `EnvSetup.model_validate(updated_actions)` to eliminate the `**` operator